gax/status_or.h

```diff
@@ -0,0 +1,214 @@
+ * `StatusOr<T>` that is not "ok". For example:
+ *
+ * @code
+ * StatusOr<Foo> foo = FetchFoo();
+ * if (!foo) {  // Same as !foo.ok()
+ *   // handle error and probably look at foo.status()
+ * } else {
+ *   foo->DoSomethingFooey();  // UB if !foo
+ * }
+ * @endcode
+ *
+ * Alternatively, you may call the `StatusOr::value()` member function.
+ *
+ * @code
+ * StatusOr<Foo> foo = FetchFoo();
+ * foo.value().DoSomethingFooey();  // May throw/crash if there is no value
+ * @endcode
+ *
+ * Functions that can fail will often return a `StatusOr<T>` instead of
+ * returning an error code and taking a `T` out-param, or rather than directly
+ * returning the `T` and throwing an exception on error. StatusOr is used so
+ * that callers can choose whether they want to explicitly check for errors,
+ * crash the program, or throw exceptions. Since constructors do not have a
+ * return value, they should be designed in such a way that they cannot fail by
+ * moving the object's complex initialization logic into a separate factory
+ * function that itself can return a `StatusOr<T>`. For example:
+ *
+ * @code
+ * class Bar {
+ *  public:
+ *   Bar(Arg arg);
+ *   ...
+ * };
+ * StatusOr<Bar> MakeBar() {
+ *   ... complicated logic that might fail
+ *   return Bar(std::move(arg));
+ * }
+ * @endcode
+ *
+ * TODO(...) - the current implementation is fairly naive with respect to `T`,
+ *   it is unlikely to work correctly for reference types, arrays, and so forth.
+ *
+ * @tparam T the type of the value.
+ */
+template<typename T>
+class StatusOr final {
+ public:
+  // There is no good definition of a default StatusOr:
+  // it can't default construct T because T may not have a default constructor,
+  // and using an unknown error code and generic message is not helpful.
+  StatusOr() = delete;
+
+  /**
+   * Creates a new `StatusOr<T>` holding the error condition @p rhs.
+   * It is undefined behavior to create a `StatusOr<T>` from an OK status.
+   *
+   * @par Post-conditions
+   * `ok() == false` and `status() == rhs`.
+   *
+   * @param rhs the status to initialize the object.
+   */
+  StatusOr(Status rhs) : status_(std::move(rhs)) {
+    if(status_.IsOk()){
+      std::cerr << "Constructing StatusOr<T> from OK status is not allowed" << std::endl;
+      std::abort();
+    }
+  }
+
+  /**
+   * Creates a new `StatusOr<T>` holding the value @p rhs.
+   *
+   * @par Post-conditions
+   * `ok() == true` and `value() == rhs`.
+   *
+   * @param rhs the value used to initialize the object.
+   */
+  StatusOr(T const& rhs) : status_(), value_(rhs) {}
+  StatusOr(T&& rhs) : status_(), value_(std::move(rhs)) {}
+
+  StatusOr(StatusOr const& rhs) : status_(rhs.status_) {
+    // Can't call ok() directly because that causes
+    // invalid memory accesses, not sure why.
+    if(status_.IsOk()) {
+      new (&value_) T(rhs.value_);
+    }
+  }
+
+  StatusOr(StatusOr&& rhs) : status_(std::move(rhs.status_)) {
+    // Can't call ok() directly because that causes
```

Comment:
This is odd and interesting. Any theories why this would happen?

Reply:
Not really. I'll look into it some more to make sure it isn't masking a more grievous error.

Reply:
Unable to reproduce in the current code. Changing to `ok()` and removing the comments.